### ISSUE TITLE
Decouple compare from test

### ISSUE THREAD

**Issue by mhavelant** [garris/BackstopJS#293]
Hi @garris ,

As far as I have seen there is no way to run compare on its own, it just runs automatically after creating the test screenshots have been finished. Is there an easy way to decouple the two?
Some backwards compatibility issues may arise from this, though. I have some ideas to fix that:
* Add an option to test --noCompare. When this is set, the test will skip the compare part
* Add a new command like testOnly (disclaimer: I'm not good at naming things)

And the compare command should have a --timestamp or something like that. This would be useful when you have 5 tests already, but you want to view older ones in the report. 

What do you think?

**Comment by garris** [comment]
Just to be sure, are you on 2.0 or 1.x?

**Comment by mhavelant** [comment]
A several-weeks-old 2.0@beta version (yes, I should update it, I know :) ).

**Comment by garris** [comment]
Can you go into the use-case for these ideas?

**Comment by mhavelant** [comment]
Yeah, sure.

Firstly, decoupling the commands will probably result in more maintainable and reusable code.

Secondly, the current way how "backstop test" works is this:
* The command creates a timestamped subfolder in the tests directory and dumps the scraped screenshots there
* Then it calls the report subcommand, which runs the comparisons, creates the reports folder with that pretty angular front end, and dumps the compare config there.

(This might be an oversimplification of the process, but this is what I see when I use the tool.)

So all this stuff is done in a single run. As stated in #281 , when you have a lot of scenarios and not that much RAM, you'll eventually run out of memory. I already encountered this, and just for fun, I monitored it wiht htop: The backstop process running 85 scenarios with 2 viewport sizes each used
* 2-300 MB of memory at the CasperJS stage
* Increased from the previous amount to 4.8 GB of usage before Linux killed it on its own.

This meant that although the screenshots remained, to comparison results were lost. Since I can't use compare on its own, I don't have a choice but to rerun test, which creates a new batch of screenshots, wasting some time.

(Note: After this little test I chopped the config up into smaller parts so at least there is this as a workaround for the memory problem.)

**TLDR**: If "backstop test" fails after taking the screenshots for some reason, you have to start again from scratch.

And thirdly, there is also what I wrote in my original post. Being able to run the compare/report part on its own would make revisiting previous test runs easier, since as far as I know, the config in the html report folder gets overwritten after each test run.

Continuing this line of thought, I think doing these following chang ideas would improve the report part of this tool:
* The decoupled test command could create a pairings.json in the current test's folder (the current timestamped folder), so this compare-older-test-runs thing would be easier.
* The new "compare" command could take these pairings.json files as arguments, and the output would be a results.json in that same (timestamped) test folder
* And to answer the "ok, but what's the point of all of this?" question: The current reports app only reads data from the config.json located in its folder. By modifying the compare command to this "output json config as result (while obviously keeping the "create diff files" functionality)", we could just tell the angular app to look through the test folder, parse the results.json files, and display these in a select field. 

This would result in an html results page which is not limited to a single config.json in its root folder, but instead, it could display every previous test run as well, making the previously mentioned revisiting of older test runs a piece of cake.

Whew, this turned out longer than originally planned. I hope everything is clear and I didn't leave out essential details needed for understanding what I meant. 

And to finish off this wall of text, I'd like to say thank you (and everyone involved) for creating this tool! My company and I are very satisfied with this tool! :)

**Comment by garris** [comment]
Thanks for the post.  There is a lot in there and I am not sure I follow all of it.   

In any case, I don't see the performance issue as being related to code organization.   I think the answer is more around throttling the number of internal async processes allowed to run at a given time.

**Comment by Patabugen** [comment]
My use-case for this is:

Right now I'm setting up Backstop and trying to get my report settings/locations correct. It'd be handy to be able to regenerate the report without waiting for the sites to load. As a workaround, I've reduced the backstop.json to just one site on one browser - but I would benefit from being able to do `backstop report` which does all the local/offline stuff usually performed after generating the test screenshots.

**Comment by rick-schubert** [comment]
@garris I would also like to be able to do this as I would want to take the screenshot with another tool since I want to be able to use different browsers. Could you perhaps just hint me to a place in the code which I would have to comment out in order to do this? Would it be that easy?
